File: height-checker/solution1051.go
// https://leetcode.com/problems/height-checker/

package main

import (
	"fmt"
	"sort"
)

// T: O(n log n), S: O(n)
func heightChecker(heights []int) int {
	expected := make([]int, len(heights))
	copy(expected, heights)
	sort.Ints(expected)
	count := 0
	for i := 0; i < len(heights); i++ {
		if heights[i] != expected[i] {
			count++
		}
	}
	return count
}

func main() {
	nums1 := []int{1, 1, 4, 2, 1, 3}
	nums2 := []int{5, 1, 2, 3, 4}
	nums3 := []int{1, 2, 3, 4, 5}
	fmt.Println(heightChecker(nums1)) // 3
	fmt.Println(heightChecker(nums2)) // 5
	fmt.Println(heightChecker(nums3)) // 0
}

File: height-checker/Solution1051.java
// https://leetcode.com/problems/height-checker/

import java.util.Arrays;

class Solution1051 {

    // T: O(n log n), S: O(n)
    public int heightChecker(int[] heights) {
        var expected = heights.clone();
        Arrays.sort(expected);
        int count = 0;
        for (int i = 0; i < heights.length; i++) {
            if (heights[i] != expected[i]) {
                count++;
            }
        }
        return count;
    }

    public static void main(String[] args) {
        var sol = new Solution1051();
        var nums1 = new int[] { 1, 1, 4, 2, 1, 3 };
        var nums2 = new int[] { 5, 1, 2, 3, 4 };
        var nums3 = new int[] { 1, 2, 3, 4, 5 };
        System.out.println(sol.heightChecker(nums1)); // 3
        System.out.println(sol.heightChecker(nums2)); // 5
        System.out.println(sol.heightChecker(nums3)); // 0
    }

}
